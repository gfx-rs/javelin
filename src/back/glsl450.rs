use crate::{
    back::glsl_common::{write_type, Error, StatementBuilder},
    Binding, BuiltIn, FastHashMap, Module, TypeInner,
};
use std::{
    fmt::{self},
    io::Write,
};

pub fn write(module: &Module, out: &mut impl Write) -> Result<(), Error> {
    writeln!(out, "#version 450 core")?;

    let mut counter = 0;
    let mut names = FastHashMap::default();

    let mut namer = |name: Option<&String>| {
        if let Some(name) = name {
            names.insert(name.clone(), ());
            name.clone()
        } else {
            counter += 1;
            while names.get(&format!("_{}", counter)).is_some() {
                counter += 1;
            }
            format!("_{}", counter)
        }
    };

    let mut structs = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let name = namer(ty.name.as_ref());
                let mut idxs = Vec::new();

                for member in members {
                    idxs.push(namer(member.name.as_ref()))
                }

                structs.insert(handle, (name, idxs));
            }
            _ => continue,
        }
    }

    // Do a second pass to build the structs
    // TODO: glsl is order dependent so we need to build structs in order
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let (name, idxs) = structs.get(&handle).unwrap();

                writeln!(out, "struct {} {{", name)?;
                for (member, name) in members.iter().zip(idxs.iter()) {
                    writeln!(
                        out,
                        "   {} {}",
                        write_type(member.ty, &module.types, &structs)?,
                        name
                    )?;
                }
                writeln!(out, "}};")?;
            }
            _ => continue,
        }
    }

    let mut globals_lookup = FastHashMap::default();

    for (handle, global) in module.global_variables.iter() {
        if let Some(Binding::BuiltIn(built_in)) = global.binding {
            let semantic = match built_in {
                BuiltIn::Position => "gl_position",
                BuiltIn::GlobalInvocationId => "gl_GlobalInvocationID",
                BuiltIn::BaseInstance => "gl_BaseInstance",
                BuiltIn::BaseVertex => "gl_BaseVertex",
                BuiltIn::ClipDistance => "gl_ClipDistance",
                BuiltIn::InstanceIndex => "gl_InstanceIndex",
                BuiltIn::VertexIndex => "gl_VertexIndex",
                BuiltIn::PointSize => "gl_PointSize",
                BuiltIn::FragCoord => "gl_FragCoord",
                BuiltIn::FrontFacing => "gl_FrontFacing",
                BuiltIn::SampleIndex => "gl_SampleID",
                BuiltIn::FragDepth => "gl_FragDepth",
                BuiltIn::LocalInvocationId => "gl_LocalInvocationID",
                BuiltIn::LocalInvocationIndex => "gl_LocalInvocationIndex",
                BuiltIn::WorkGroupId => "gl_WorkGroupID",
            };

            globals_lookup.insert(handle, String::from(semantic));
            continue;
        }

        if let Some(ref binding) = global.binding {
            write!(out, "layout({}) ", binding.writer())?;
        }

        let name = namer(global.name.as_ref());

        writeln!(
            out,
            "{}{} {};",
            global.class.write_glsl(),
            write_type(global.ty, &module.types, &structs)?,
            name
        )?;

        globals_lookup.insert(handle, name);
    }

    let mut functions = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, func) in module.functions.iter() {
        functions.insert(handle, namer(func.name.as_ref()));
    }

    let mut types = module.types.clone();
    let mut typifier = crate::proc::Typifier::new();

    // TODO: glsl is order dependent so we need to build functions in order
    for (handle, func) in module.functions.iter() {
        let name = functions.get(&handle).unwrap();

        writeln!(
            out,
            "{} {}({}) {{",
            func.return_type
                .map_or(Ok(String::from("void")), |ty| write_type(
                    ty, &types, &structs
                ))?,
            name,
            func.parameter_types
                .iter()
                .map(|ty| write_type(*ty, &types, &structs))
                .collect::<Result<Vec<_>, _>>()?
                .join(","),
        )?;

        let mut builder = StatementBuilder {
            functions: &functions,
            globals: &globals_lookup,
            locals_lookup: &func
                .local_variables
                .iter()
                .map(|(handle, local)| (handle, namer(local.name.as_ref())))
                .collect(),
            structs: &structs,
            args: &func
                .parameter_types
                .iter()
                .enumerate()
                .map(|(pos, _)| (pos as u32, namer(None)))
                .collect(),
            expressions: &func.expressions,
            types: &mut types,
            locals: &func.local_variables,
            typifier: &mut typifier,
        };

        for sta in func.body.iter() {
            writeln!(out, "{}", sta.write_glsl(module, &mut builder)?)?;
        }

        writeln!(out, "}}")?;
    }

    Ok(())
}

struct BindingWriter<'a> {
    inner: &'a Binding,
}

impl Binding {
    fn writer<'a>(&'a self) -> BindingWriter<'a> {
        BindingWriter { inner: self }
    }
}

impl<'a> fmt::Display for BindingWriter<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.inner {
            Binding::BuiltIn(_) => write!(f, ""), // Ignore because they are variables with a predefined name
            Binding::Location(location) => write!(f, "location={}", location),
            Binding::Descriptor { set, binding } => write!(f, "set={},binding={}", set, binding),
        }
    }
}
