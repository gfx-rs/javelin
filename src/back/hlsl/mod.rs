mod keywords;
mod writer;

use std::fmt::Error as FmtError;
use thiserror::Error;

pub use writer::Writer;

#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
pub struct ShaderModel(u16);

impl Default for ShaderModel {
    fn default() -> Self {
        ShaderModel(50)
    }
}

#[derive(Error, Debug)]
pub enum Error {
    #[error(transparent)]
    IoError(#[from] FmtError),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedShaderModel(ShaderModel),
    #[error("A scalar with an unsupported width was requested: {0:?} {1:?}")]
    UnsupportedScalar(crate::ScalarKind, crate::Bytes),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedBuiltIn(crate::BuiltIn),
    #[error("{0}")]
    Unimplemented(String), // TODO: Error used only during development
}

pub fn write_string(
    module: &crate::Module,
    info: &crate::valid::ModuleInfo,
    shader_model: ShaderModel,
) -> Result<String, Error> {
    let mut w = Writer::new(String::new(), shader_model);
    w.write(module, info)?;
    let output = w.finish();
    Ok(output)
}
