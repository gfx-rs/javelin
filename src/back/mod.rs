//! Functions which export shader modules into binary and text formats.

#[cfg(feature = "glsl-out")]
pub mod glsl;
#[cfg(feature = "msl-out")]
pub mod msl;
#[cfg(feature = "spv-out")]
pub mod spv;

impl crate::Expression {
    /// Returns the ref count, upon reaching which this expression
    /// should be considered for baking.
    #[allow(dead_code)]
    fn bake_ref_count(&self) -> usize {
        match *self {
            // accesses are never cached, only loads are
            crate::Expression::Access { .. } | crate::Expression::AccessIndex { .. } => !0,
            // image operations look better when isolated
            crate::Expression::ImageSample { .. } | crate::Expression::ImageLoad { .. } => 1,
            // cache expressions that are referenced multiple times
            _ => 2,
        }
    }
}
