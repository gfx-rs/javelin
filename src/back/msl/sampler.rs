#[derive(Clone, Debug, PartialEq)]
pub enum Coord {
    Normalized,
    Pixel,
}

impl Default for Coord {
    fn default() -> Self {
        Self::Normalized
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Address {
    Repeat,
    MirroredRepeat,
    ClampToEdge,
    ClampToZero,
    ClampToBorder,
}

impl Default for Address {
    fn default() -> Self {
        Self::ClampToEdge
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum BorderColor {
    TransparentBlack,
    OpaqueBlack,
    OpaqueWhite,
}

impl Default for BorderColor {
    fn default() -> Self {
        Self::TransparentBlack
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Filter {
    Nearest,
    Linear,
}

impl Default for Filter {
    fn default() -> Self {
        Self::Nearest
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum CompareFunc {
    Never,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    Always,
}

impl Default for CompareFunc {
    fn default() -> Self {
        Self::Never
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct InlineSampler {
    pub coord: Coord,
    pub address: [Address; 3],
    pub border_color: BorderColor,
    pub mag_filter: Filter,
    pub min_filter: Filter,
    pub mip_filter: Option<Filter>,
    pub compare_func: CompareFunc,
}
