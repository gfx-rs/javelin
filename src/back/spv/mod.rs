mod layout;
pub mod parser;

use spirv::*;

struct Module {
    physical_layout: PhysicalLayout,
    logical_layout: LogicalLayout,
}

impl Module {
    fn new(header: &crate::Header) -> Self {
        Module {
            physical_layout: PhysicalLayout::new(header, 0x0u32),
            logical_layout: LogicalLayout::new(),
        }
    }
}

struct PhysicalLayout {
    magic_number: Word,
    version: Word,
    generator: Word,
    bound: Word,
    instruction_schema: Word,
}

impl PhysicalLayout {
    fn new(header: &crate::Header, instruction_schema: Word) -> Self {
        let version: Word = (0x0u32 << 24)
            | ((header.version.0 as u32) << 16)
            | ((header.version.1 as u32) << 8)
            | header.version.2 as u32;

        PhysicalLayout {
            magic_number: MAGIC_NUMBER,
            version,
            generator: header.generator,
            bound: 0,
            instruction_schema,
        }
    }

    fn in_words(&self) -> Vec<Word> {
        let mut words: Vec<Word> = vec![];

        words.push(self.magic_number);
        words.push(self.version);
        words.push(self.generator);

        match self.bound {
            0 => panic!("Bound cannot be 0"),
            _ => words.push(self.bound),
        }

        words.push(self.instruction_schema);
        words
    }
}

pub(crate) struct LogicalLayout {
    capabilities: Vec<Word>,
    extensions: Vec<Word>,
    ext_inst_imports: Vec<Word>,
    memory_model: Vec<Word>,
    entry_points: Vec<Word>,
    execution_modes: Vec<Word>,
    debugs: Vec<Word>,
    annotations: Vec<Word>,
    type_declarations: Vec<Word>,
    constants: Vec<Word>,
    global_variables: Vec<Word>,
    function_declarations: Vec<Word>,
    function_definitions: Vec<Word>,
}

impl LogicalLayout {
    fn new() -> Self {
        LogicalLayout {
            capabilities: vec![],
            extensions: vec![],
            ext_inst_imports: vec![],
            memory_model: vec![],
            entry_points: vec![],
            execution_modes: vec![],
            debugs: vec![],
            annotations: vec![],
            type_declarations: vec![],
            constants: vec![],
            global_variables: vec![],
            function_declarations: vec![],
            function_definitions: vec![],
        }
    }

    fn in_words(&self) -> Vec<u32> {
        let mut words: Vec<Word> = vec![];

        for capability in self.capabilities.iter() {
            words.push(*capability);
        }

        for extension in self.extensions.iter() {
            words.push(*extension);
        }

        for ext_inst_import in self.ext_inst_imports.iter() {
            words.push(*ext_inst_import);
        }

        for memory_model in self.memory_model.iter() {
            words.push(*memory_model);
        }

        for entry_point in self.entry_points.iter() {
            words.push(*entry_point);
        }

        for execution_mode in self.execution_modes.iter() {
            words.push(*execution_mode);
        }

        for debug in self.debugs.iter() {
            words.push(*debug);
        }

        for annotation in self.annotations.iter() {
            words.push(*annotation);
        }

        for type_declaration in self.type_declarations.iter() {
            words.push(*type_declaration);
        }

        for constants in self.constants.iter() {
            words.push(*constants);
        }

        for global_variables in self.global_variables.iter() {
            words.push(*global_variables);
        }

        for function_declaration in self.function_declarations.iter() {
            words.push(*function_declaration);
        }

        for function_definition in self.function_definitions.iter() {
            words.push(*function_definition);
        }

        words
    }
}

struct Instruction {
    op: Op,
    wc: u32,
    type_id: Option<Word>,
    result_id: Option<Word>,
    operands: Vec<Word>,
}

impl Instruction {
    fn new(op: Op) -> Self {
        Instruction {
            op,
            wc: 1, // Always start at 1 for the first word (OP + WC),
            type_id: None,
            result_id: None,
            operands: vec![],
        }
    }

    fn set_type(&mut self, id: Word) {
        assert!(self.type_id.is_none(), "Type can only be set once");
        self.type_id = Some(id);
        self.wc += 1;
    }

    fn set_result(&mut self, id: Word) {
        assert!(self.result_id.is_none(), "Result can only be set once");
        self.result_id = Some(id);
        self.wc += 1;
    }

    fn add_operand(&mut self, operand: Word) {
        self.operands.push(operand);
        self.wc += 1;
    }

    fn add_operands(&mut self, operands: Vec<Word>) {
        for operand in operands {
            self.add_operand(operand);
        }
    }

    fn to_words(&self) -> Vec<Word> {
        let mut words = Vec::with_capacity(self.wc as usize);

        let wc_op = (self.wc << 16 | self.op as u32) as u32;

        words.push(wc_op);

        if self.type_id.is_some() {
            words.push(self.type_id.unwrap());
        }

        if self.result_id.is_some() {
            words.push(self.result_id.unwrap());
        }

        for operand in self.operands.iter() {
            words.push(*operand);
        }

        words
    }
}
