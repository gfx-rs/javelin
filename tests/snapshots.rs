use serde::Deserialize;

#[derive(Hash, PartialEq, Eq, Deserialize)]
enum Stage {
    Vertex,
    Fragment,
    Compute,
}

#[derive(Hash, PartialEq, Eq, Deserialize)]
struct BindSource {
    stage: Stage,
    group: u32,
    binding: u32,
}

#[derive(Deserialize)]
struct BindTarget {
    #[serde(default)]
    buffer: Option<u8>,
    #[serde(default)]
    texture: Option<u8>,
    #[serde(default)]
    sampler: Option<u8>,
    #[serde(default)]
    mutable: bool,
}

#[derive(Default, Deserialize)]
struct Parameters {
    #[serde(default)]
    spv_flow_dump_prefix: String,
    #[cfg_attr(not(feature = "spv-out"), allow(dead_code))]
    spv_capabilities: naga::FastHashSet<spirv::Capability>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    mtl_bindings: naga::FastHashMap<BindSource, BindTarget>,
}

#[cfg(feature = "msl-out")]
fn check_output_msl(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::msl;
    let mut binding_map = msl::BindingMap::default();
    for (key, value) in params.mtl_bindings.iter() {
        binding_map.insert(
            msl::BindSource {
                stage: match key.stage {
                    Stage::Vertex => naga::ShaderStage::Vertex,
                    Stage::Fragment => naga::ShaderStage::Fragment,
                    Stage::Compute => naga::ShaderStage::Compute,
                },
                group: key.group,
                binding: key.binding,
            },
            msl::BindTarget {
                buffer: value.buffer,
                texture: value.texture,
                sampler: value.sampler,
                mutable: value.mutable,
            },
        );
    }
    let options = msl::Options {
        lang_version: (1, 0),
        spirv_cross_compatibility: false,
        binding_map,
    };
    let (msl, _) = msl::write_string(&module, &options).unwrap();
    insta::assert_snapshot!(format!("{}{}", name, ".msl"), msl);
}

#[cfg(feature = "spv-out")]
fn check_output_spv(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::spv;
    use rspirv::binary::Disassemble;
    let spv = spv::write_vec(
        &module,
        spv::WriterFlags::NONE,
        params.spv_capabilities.clone(),
    )
    .unwrap();
    let dis = rspirv::dr::load_words(spv)
        .expect("Produced invalid SPIR-V")
        .disassemble();
    insta::assert_snapshot!(format!("{}{}", name, ".spvasm"), dis);
}

#[cfg(feature = "wgsl-in")]
fn convert_wgsl(name: &str) {
    let params =
        match std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".param.ron")) {
            Ok(string) => ron::de::from_str(&string).expect("Couldn't find param file"),
            Err(_) => Parameters::default(),
        };

    let module = naga::front::wgsl::parse_str(
        &std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".wgsl"))
            .expect("Couldn't find wgsl file"),
    )
    .unwrap();
    naga::proc::Validator::new().validate(&module).unwrap();

    #[cfg(feature = "msl-out")]
    check_output_msl(&module, name, &params);
    #[cfg(feature = "spv-out")]
    check_output_spv(&module, name, &params);
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_quad() {
    convert_wgsl("quad");
}
