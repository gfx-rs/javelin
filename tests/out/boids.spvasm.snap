---
source: tests/snapshots.rs
expression: dis
---
; SPIR-V
; Version: 1.0
; Generator: rspirv
; Bound: 221
OpCapability Shader
%1 = OpExtInstImport "GLSL.std.450"
OpMemoryModel Logical GLSL450
OpEntryPoint GLCompute %41 "main" %25
OpExecutionMode %41 LocalSize 64 1 1
OpSource GLSL 450
OpName %3 "NUM_PARTICLES"
OpName %16 "SimParams"
OpMemberName %16 0 "deltaT"
OpMemberName %16 1 "rule1Distance"
OpMemberName %16 2 "rule2Distance"
OpMemberName %16 3 "rule3Distance"
OpMemberName %16 4 "rule1Scale"
OpMemberName %16 5 "rule2Scale"
OpMemberName %16 6 "rule3Scale"
OpName %15 "params"
OpName %19 "Particles"
OpMemberName %19 0 "particles"
OpName %21 "Particle"
OpMemberName %21 0 "pos"
OpMemberName %21 1 "vel"
OpName %18 "particlesSrc"
OpName %24 "particlesDst"
OpName %25 "gl_GlobalInvocationID"
OpName %28 "vPos"
OpName %30 "vVel"
OpName %31 "cMass"
OpName %32 "cVel"
OpName %33 "colVel"
OpName %34 "cMassCount"
OpName %36 "cVelCount"
OpName %37 "pos"
OpName %38 "vel"
OpName %39 "i"
OpName %41 "main"
OpName %41 "main"
OpDecorate %16 Block
OpMemberDecorate %16 0 Offset 0
OpMemberDecorate %16 1 Offset 4
OpMemberDecorate %16 2 Offset 8
OpMemberDecorate %16 3 Offset 12
OpMemberDecorate %16 4 Offset 16
OpMemberDecorate %16 5 Offset 20
OpMemberDecorate %16 6 Offset 24
OpDecorate %15 DescriptorSet 0
OpDecorate %15 Binding 0
OpDecorate %19 BufferBlock
OpMemberDecorate %19 0 Offset 0
OpDecorate %20 ArrayStride 16
OpMemberDecorate %21 0 Offset 0
OpMemberDecorate %21 1 Offset 8
OpDecorate %18 NonWritable
OpDecorate %18 DescriptorSet 0
OpDecorate %18 Binding 1
OpDecorate %24 DescriptorSet 0
OpDecorate %24 Binding 2
OpDecorate %25 BuiltIn GlobalInvocationId
%2 = OpTypeVoid
%4 = OpTypeInt 32 1
%3 = OpConstant  %4  1500
%6 = OpTypeFloat 32
%5 = OpConstant  %6  0.0
%7 = OpConstant  %4  0
%9 = OpTypeInt 32 0
%8 = OpConstant  %9  0
%10 = OpConstant  %4  1
%11 = OpConstant  %9  1
%12 = OpConstant  %6  1.0
%13 = OpConstant  %6  0.1
%14 = OpConstant  %6  -1.0
%16 = OpTypeStruct %6 %6 %6 %6 %6 %6 %6
%17 = OpTypePointer Uniform %16
%15 = OpVariable  %17  Uniform
%22 = OpTypeVector %6 2
%21 = OpTypeStruct %22 %22
%20 = OpTypeRuntimeArray %21
%19 = OpTypeStruct %20
%23 = OpTypePointer Uniform %19
%18 = OpVariable  %23  Uniform
%24 = OpVariable  %23  Uniform
%26 = OpTypeVector %9 3
%27 = OpTypePointer Input %26
%25 = OpVariable  %27  Input
%29 = OpTypePointer Function %22
%35 = OpTypePointer Function %4
%40 = OpTypePointer Function %9
%42 = OpTypeFunction %2
%47 = OpTypeBool
%51 = OpTypePointer Uniform %20
%52 = OpTypePointer Uniform %21
%53 = OpTypePointer Uniform %22
%54 = OpConstant  %4  0
%55 = OpConstant  %4  0
%58 = OpConstant  %4  1
%59 = OpConstant  %4  0
%78 = OpConstant  %4  0
%79 = OpConstant  %4  0
%83 = OpConstant  %4  1
%84 = OpConstant  %4  0
%90 = OpTypePointer Uniform %6
%91 = OpConstant  %4  1
%105 = OpConstant  %4  2
%119 = OpConstant  %4  3
%154 = OpConstant  %4  4
%160 = OpConstant  %4  5
%166 = OpConstant  %4  6
%179 = OpConstant  %4  0
%189 = OpTypePointer Function %6
%190 = OpConstant  %4  0
%197 = OpConstant  %4  0
%204 = OpConstant  %4  1
%211 = OpConstant  %4  1
%214 = OpConstant  %4  0
%215 = OpConstant  %4  0
%218 = OpConstant  %4  1
%219 = OpConstant  %4  0
%41 = OpFunction  %2  None %42
%43 = OpLabel
%39 = OpVariable  %40  Function %8
%36 = OpVariable  %35  Function %7
%32 = OpVariable  %29  Function
%28 = OpVariable  %29  Function
%37 = OpVariable  %29  Function
%33 = OpVariable  %29  Function
%30 = OpVariable  %29  Function
%38 = OpVariable  %29  Function
%34 = OpVariable  %35  Function %7
%31 = OpVariable  %29  Function
OpBranch %44
%44 = OpLabel
%45 = OpLoad  %26  %25
%46 = OpCompositeExtract  %9  %45 0
%48 = OpUGreaterThanEqual  %47  %46 %3
OpSelectionMerge %49 None
OpBranchConditional %48 %50 %49
%50 = OpLabel
OpReturn
%49 = OpLabel
%56 = OpAccessChain  %53  %18 %55 %46 %54
%57 = OpLoad  %22  %56
OpStore %28 %57
%60 = OpAccessChain  %53  %18 %59 %46 %58
%61 = OpLoad  %22  %60
OpStore %30 %61
%62 = OpCompositeConstruct  %22  %5 %5
OpStore %31 %62
%63 = OpCompositeConstruct  %22  %5 %5
OpStore %32 %63
%64 = OpCompositeConstruct  %22  %5 %5
OpStore %33 %64
OpBranch %65
%65 = OpLabel
OpLoopMerge %66 %68 None
OpBranch %67
%67 = OpLabel
%69 = OpLoad  %9  %39
%70 = OpUGreaterThanEqual  %47  %69 %3
OpSelectionMerge %71 None
OpBranchConditional %70 %72 %71
%72 = OpLabel
OpBranch %66
%71 = OpLabel
%73 = OpLoad  %9  %39
%74 = OpIEqual  %47  %73 %46
OpSelectionMerge %75 None
OpBranchConditional %74 %76 %75
%76 = OpLabel
OpBranch %68
%75 = OpLabel
%77 = OpLoad  %9  %39
%80 = OpAccessChain  %53  %18 %79 %77 %78
%81 = OpLoad  %22  %80
OpStore %37 %81
%82 = OpLoad  %9  %39
%85 = OpAccessChain  %53  %18 %84 %82 %83
%86 = OpLoad  %22  %85
OpStore %38 %86
%87 = OpLoad  %22  %37
%88 = OpLoad  %22  %28
%89 = OpExtInst  %6  %1 Distance %87 %88
%92 = OpAccessChain  %90  %15 %91
%93 = OpLoad  %6  %92
%94 = OpFOrdLessThan  %47  %89 %93
OpSelectionMerge %95 None
OpBranchConditional %94 %96 %95
%96 = OpLabel
%97 = OpLoad  %22  %31
%98 = OpLoad  %22  %37
%99 = OpFAdd  %22  %97 %98
OpStore %31 %99
%100 = OpLoad  %4  %34
%101 = OpIAdd  %4  %100 %10
OpStore %34 %101
OpBranch %95
%95 = OpLabel
%102 = OpLoad  %22  %37
%103 = OpLoad  %22  %28
%104 = OpExtInst  %6  %1 Distance %102 %103
%106 = OpAccessChain  %90  %15 %105
%107 = OpLoad  %6  %106
%108 = OpFOrdLessThan  %47  %104 %107
OpSelectionMerge %109 None
OpBranchConditional %108 %110 %109
%110 = OpLabel
%111 = OpLoad  %22  %33
%112 = OpLoad  %22  %37
%113 = OpLoad  %22  %28
%114 = OpFSub  %22  %112 %113
%115 = OpFSub  %22  %111 %114
OpStore %33 %115
OpBranch %109
%109 = OpLabel
%116 = OpLoad  %22  %37
%117 = OpLoad  %22  %28
%118 = OpExtInst  %6  %1 Distance %116 %117
%120 = OpAccessChain  %90  %15 %119
%121 = OpLoad  %6  %120
%122 = OpFOrdLessThan  %47  %118 %121
OpSelectionMerge %123 None
OpBranchConditional %122 %124 %123
%124 = OpLabel
%125 = OpLoad  %22  %32
%126 = OpLoad  %22  %38
%127 = OpFAdd  %22  %125 %126
OpStore %32 %127
%128 = OpLoad  %4  %36
%129 = OpIAdd  %4  %128 %10
OpStore %36 %129
OpBranch %123
%123 = OpLabel
OpBranch %68
%68 = OpLabel
%130 = OpLoad  %9  %39
%131 = OpIAdd  %9  %130 %11
OpStore %39 %131
OpBranch %65
%66 = OpLabel
%132 = OpLoad  %4  %34
%133 = OpSGreaterThan  %47  %132 %7
OpSelectionMerge %134 None
OpBranchConditional %133 %135 %134
%135 = OpLabel
%136 = OpLoad  %22  %31
%137 = OpLoad  %4  %34
%138 = OpConvertSToF  %6  %137
%139 = OpFDiv  %6  %12 %138
%140 = OpVectorTimesScalar  %22  %136 %139
%141 = OpLoad  %22  %28
%142 = OpFSub  %22  %140 %141
OpStore %31 %142
OpBranch %134
%134 = OpLabel
%143 = OpLoad  %4  %36
%144 = OpSGreaterThan  %47  %143 %7
OpSelectionMerge %145 None
OpBranchConditional %144 %146 %145
%146 = OpLabel
%147 = OpLoad  %22  %32
%148 = OpLoad  %4  %36
%149 = OpConvertSToF  %6  %148
%150 = OpFDiv  %6  %12 %149
%151 = OpVectorTimesScalar  %22  %147 %150
OpStore %32 %151
OpBranch %145
%145 = OpLabel
%152 = OpLoad  %22  %30
%153 = OpLoad  %22  %31
%155 = OpAccessChain  %90  %15 %154
%156 = OpLoad  %6  %155
%157 = OpVectorTimesScalar  %22  %153 %156
%158 = OpFAdd  %22  %152 %157
%159 = OpLoad  %22  %33
%161 = OpAccessChain  %90  %15 %160
%162 = OpLoad  %6  %161
%163 = OpVectorTimesScalar  %22  %159 %162
%164 = OpFAdd  %22  %158 %163
%165 = OpLoad  %22  %32
%167 = OpAccessChain  %90  %15 %166
%168 = OpLoad  %6  %167
%169 = OpVectorTimesScalar  %22  %165 %168
%170 = OpFAdd  %22  %164 %169
OpStore %30 %170
%171 = OpLoad  %22  %30
%172 = OpExtInst  %22  %1 Normalize %171
%173 = OpLoad  %22  %30
%174 = OpExtInst  %6  %1 Length %173
%175 = OpExtInst  %6  %1 FClamp %174 %5 %13
%176 = OpVectorTimesScalar  %22  %172 %175
OpStore %30 %176
%177 = OpLoad  %22  %28
%178 = OpLoad  %22  %30
%180 = OpAccessChain  %90  %15 %179
%181 = OpLoad  %6  %180
%182 = OpVectorTimesScalar  %22  %178 %181
%183 = OpFAdd  %22  %177 %182
OpStore %28 %183
%184 = OpLoad  %22  %28
%185 = OpCompositeExtract  %6  %184 0
%186 = OpFOrdLessThan  %47  %185 %14
OpSelectionMerge %187 None
OpBranchConditional %186 %188 %187
%188 = OpLabel
%191 = OpAccessChain  %189  %28 %190
OpStore %191 %12
OpBranch %187
%187 = OpLabel
%192 = OpLoad  %22  %28
%193 = OpCompositeExtract  %6  %192 0
%194 = OpFOrdGreaterThan  %47  %193 %12
OpSelectionMerge %195 None
OpBranchConditional %194 %196 %195
%196 = OpLabel
%198 = OpAccessChain  %189  %28 %197
OpStore %198 %14
OpBranch %195
%195 = OpLabel
%199 = OpLoad  %22  %28
%200 = OpCompositeExtract  %6  %199 1
%201 = OpFOrdLessThan  %47  %200 %14
OpSelectionMerge %202 None
OpBranchConditional %201 %203 %202
%203 = OpLabel
%205 = OpAccessChain  %189  %28 %204
OpStore %205 %12
OpBranch %202
%202 = OpLabel
%206 = OpLoad  %22  %28
%207 = OpCompositeExtract  %6  %206 1
%208 = OpFOrdGreaterThan  %47  %207 %12
OpSelectionMerge %209 None
OpBranchConditional %208 %210 %209
%210 = OpLabel
%212 = OpAccessChain  %189  %28 %211
OpStore %212 %14
OpBranch %209
%209 = OpLabel
%213 = OpLoad  %22  %28
%216 = OpAccessChain  %53  %24 %215 %46 %214
OpStore %216 %213
%217 = OpLoad  %22  %30
%220 = OpAccessChain  %53  %24 %219 %46 %218
OpStore %220 %217
OpReturn
OpFunctionEnd
